### PR TITLE
Feature/automatic text/store feedback consistency

### PR DESCRIPTION
**Pull request by bigultekin**
⚠️ Depends on https://github.com/ls1intum/Athene/pull/18 to be deployed as well!

---

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) ~on the test server https://artemistest.ase.in.tum.de.~
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] ~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] ~Client: I added multiple integration tests (Jest) related to the features~
- [x] ~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~
- [x] ~Client: I documented the TypeScript code using JSDoc style.~
- [x] ~Client: I added multiple screenshots/screencasts of my UI changes~
- [x] ~Client: I translated all the newly inserted strings into German and English~

### Motivation and Context

As mentioned in the feature request [#1877 ](https://github.com/ls1intum/Artemis/issues/1877), we want to store feedback conflicts that is found during the assessment of text exercises. 
The conflicts are found with an Athene service. Corresponding PR in Athene https://github.com/ls1intum/Athene/pull/18

### Description

Workflow:

1.  When a text assessment is submitted, an asynchronous call is made to the remote service only if the automatic assessment is enabled and assessed text blocks are already assigned to a cluster. 
2. Feedback consistency is checked in Athene (≈1-2 seconds)
3. Found conflicts are returned and saved 

Changes:
- TextAssessmentConflictService is added to communicate with Athene service
- Added TextAssessmentConflict entity and corresponding repository
- Minor Feedback entity extension
- Added multiple test cases
   - Added TextAssessmentConflictServiceTest to test the communication and response of the Athene
   - Added AutomaticTextAssessmentConflictServiceTest to mock the response of the Athene and to test the database operations

### Steps for Testing

Testing in the test server is not possible since https://github.com/ls1intum/Athene/pull/18 is not deployed yet. So, please test it locally. 

1. Start Athene with this [branch](https://github.com/ls1intum/Athene/tree/feature/feedback-consistency)
2. Log in to Artemis
3. Create a text exercise with automatic assessments are enabled
4. Submit a couple of text submissions (use similar texts so they can be clustered together)
5. Wait for the submission due date so Athene can cluster text blocks (you can check Artemis logs or docker dashboard for Clustering component in Athene to see if clustering is started/done.)
6. Assess multiple submissions (with completely different feedback comments and scores 🙂 )
7. Check the database if the inconsistencies are stored 

### Test Coverage

- AutomaticTextAssessmentConflictService - 91%
- TextAssessmentConflictService - 37% ( connector class - test is skipped because Athene is not updated yet.)

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 1]
```diff
@@ -207,6 +215,22 @@ public void setGradingInstruction(GradingInstruction gradingInstruction) {
...
+
+    public void setFirst_conflicts(List<TextAssessmentConflict> first_conflicts) {
+        this.first_conflicts = first_conflicts;
+    }
+
+    public List<TextAssessmentConflict> getSecond_conflicts() {
+        return second_conflicts;
+    }
+
+    public void setSecond_conflicts(List<TextAssessmentConflict> second_conflicts) {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:230]: Codacy found an issue: [The instance method name 'setSecond_conflicts' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
+package de.tum.in.www1.artemis.domain;
+
+import java.io.Serializable;
+import java.time.ZonedDateTime;
+
+import javax.persistence.*;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java:6]: Codacy found an issue: [Avoid unused imports such as 'javax.persistence'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/TextAssessmentConflictService.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        public List<TextAssessmentConflictResponseDTO> feedbackInconsistencies;
+    }
+    // endregion
+
+    @Value("${artemis.automatic-text.feedback-consistency-url}")
+    private String API_ENDPOINT;
+
+    @Value("${artemis.automatic-text.secret}")
+    private String API_SECRET;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/TextAssessmentConflictService.java:46]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 4]
```diff
@@ -207,6 +215,22 @@ public void setGradingInstruction(GradingInstruction gradingInstruction) {
...
 
+    public List<TextAssessmentConflict> getFirst_conflicts() {
+        return first_conflicts;
+    }
+
+    public void setFirst_conflicts(List<TextAssessmentConflict> first_conflicts) {
+        this.first_conflicts = first_conflicts;
+    }
+
+    public List<TextAssessmentConflict> getSecond_conflicts() {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:226]: Codacy found an issue: [The instance method name 'getSecond_conflicts' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    public void setCreated_at(ZonedDateTime created_at) {
+        this.created_at = created_at;
+    }
+
+    public ZonedDateTime getSolved_at() {
+        return solved_at;
+    }
+
+    public void setSolved_at(ZonedDateTime solved_at) {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java:72]: Codacy found an issue: [The instance method name 'setSolved_at' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    public void setConflict(Boolean conflict) {
+        this.conflict = conflict;
+    }
+
+    public ZonedDateTime getCreated_at() {
+        return created_at;
+    }
+
+    public void setCreated_at(ZonedDateTime created_at) {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java:64]: Codacy found an issue: [The instance method name 'setCreated_at' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/TextAssessmentConflictService.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    private static class Response {
+
+        public List<TextAssessmentConflictResponseDTO> feedbackInconsistencies;
+    }
+    // endregion
+
+    @Value("${artemis.automatic-text.feedback-consistency-url}")
+    private String API_ENDPOINT;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/TextAssessmentConflictService.java:43]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    public Boolean getConflict() {
+        return conflict;
+    }
+
+    public void setConflict(Boolean conflict) {
+        this.conflict = conflict;
+    }
+
+    public ZonedDateTime getCreated_at() {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java:60]: Codacy found an issue: [The instance method name 'getCreated_at' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/service/AutomaticTextAssessmentConflictService.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import de.tum.in.www1.artemis.repository.TextAssessmentConflictRepository;
+import de.tum.in.www1.artemis.service.connectors.TextAssessmentConflictService;
+import de.tum.in.www1.artemis.service.dto.TextAssessmentConflictRequestDTO;
+import de.tum.in.www1.artemis.service.dto.TextAssessmentConflictResponseDTO;
+
+@Service
+@Profile("automaticText")
+public class AutomaticTextAssessmentConflictService {
+
+    private final Logger log = LoggerFactory.getLogger(AutomaticTextAssessmentConflictService.class);
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/AutomaticTextAssessmentConflictService.java:30]: Codacy found an issue: [Avoid unused private fields such as 'log'.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        return created_at;
+    }
+
+    public void setCreated_at(ZonedDateTime created_at) {
+        this.created_at = created_at;
+    }
+
+    public ZonedDateTime getSolved_at() {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/TextAssessmentConflict.java:68]: Codacy found an issue: [The instance method name 'getSolved_at' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/TextAssessmentConflictService.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Value("${artemis.automatic-text.secret}")
+    private String API_SECRET;
+
+    private RemoteArtemisServiceConnector<Request, Response> connector = new RemoteArtemisServiceConnector<>(log, Response.class);
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/TextAssessmentConflictService.java:48]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 12]
```diff
@@ -207,6 +215,22 @@ public void setGradingInstruction(GradingInstruction gradingInstruction) {
         this.gradingInstruction = gradingInstruction;
     }
 
+    public List<TextAssessmentConflict> getFirst_conflicts() {
+        return first_conflicts;
+    }
+
+    public void setFirst_conflicts(List<TextAssessmentConflict> first_conflicts) {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:222]: Codacy found an issue: [The instance method name 'setFirst_conflicts' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 13]
```diff
@@ -207,6 +215,22 @@ public void setGradingInstruction(GradingInstruction gradingInstruction) {
         this.gradingInstruction = gradingInstruction;
     }
 
+    public List<TextAssessmentConflict> getFirst_conflicts() {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:218]: Codacy found an issue: [The instance method name 'getFirst_conflicts' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968856)

**Comment by krusche** [comment]
Please make sure that your tests cover all important cases, in particular that the deletion of courses, exercises, participations, submissions and results is always possible. This has to be covered in tests.

Also make sure to add the test coverage (as required in the new PR template) to the PR description

**Review by CanArisan — APPROVED** [review]
Followed the testing steps and inconsistencies were stored in the database 👍
The stored inconsistency looked like this:
image

**Comment by bigultekin** [comment]
> Followed the testing steps and inconsistencies were stored in the database 👍
> The stored inconsistency looked like this:
> image

Thank you @CanArisan for testing!

**Pull request merged**